src/job.rs

```diff
@@ -257,6 +255,130 @@ impl Job {
+                        let (job_id, params) = row.ok()?;
+                        let msg_id = params.get_msg_id()?;
+                        Some((job_id, msg_id))
+                    })
+                    .collect();
+                Ok(res)
+            },
+        )?;
+
+        // Load corresponding RFC724 message IDs
+        let mut job_ids = Vec::new();
+        let mut rfc724_mids = Vec::new();
+        for (job_id, msg_id) in res {
+            if let Ok(Message { rfc724_mid, .. }) = Message::load_from_db(context, msg_id) {
+                job_ids.push(job_id);
+                rfc724_mids.push(rfc724_mid);
+            }
+        }
+        Ok((job_ids, rfc724_mids))
+    }
+
+    #[allow(non_snake_case)]
+    fn SendMdn(&mut self, context: &Context) -> Status {
+        if !context.get_config_bool(Config::MdnsEnabled) {
+            // User has disabled MDNs after job scheduling but before
+            // execution.
+            return Status::Finished(Err(format_err!("MDNs are disabled")));
+        }
+
+        let contact_id = self.foreign_id;
+        let contact = job_try!(Contact::load_from_db(context, contact_id));
+        if contact.is_blocked() {
+            return Status::Finished(Err(format_err!("Contact is blocked")));
+        }
+
+        let msg_id = if let Some(msg_id) = self.param.get_msg_id() {
+            msg_id
+        } else {
+            return Status::Finished(Err(format_err!(
+                "SendMdn job has invalid parameters: {}",
+                self.param
+            )));
+        };
+
+        // Try to aggregate other SendMdn jobs and send a combined MDN.
+        let (additional_job_ids, additional_rfc724_mids) = self
+            .get_additional_mdn_jobs(context, contact_id)
+            .unwrap_or_default();
+
+        if !additional_rfc724_mids.is_empty() {
+            info!(
+                context,
+                "SendMdn job: aggregating {} additional MDNs",
+                additional_rfc724_mids.len()
+            )
+        }
+
+        let msg = job_try!(Message::load_from_db(context, msg_id));
+        let mimefactory = job_try!(MimeFactory::from_mdn(context, &msg, additional_rfc724_mids));
+        let rendered_msg = job_try!(mimefactory.render());
+        let body = rendered_msg.message;
+
+        let addr = contact.get_addr();
+        let recipient = job_try!(async_smtp::EmailAddress::new(addr.to_string())
+            .map_err(|err| format_err!("invalid recipient: {} {:?}", addr, err)));
+        let recipients = vec![recipient];
+
+        /* connect to SMTP server, if not yet done */
+        if !context.smtp.lock().unwrap().is_connected() {
+            let loginparam = LoginParam::from_database(context, "configured_");
+            if let Err(err) = context.smtp.lock().unwrap().connect(context, &loginparam) {
+                warn!(context, "SMTP connection failure: {:?}", err);
+                return Status::RetryLater;
+            }
+        }
```

Comment:
pelase factor this code paragraph out, it's exactly the same as in "fn SendMsgToSmtp".